monix-eval/shared/src/main/scala/monix/eval/internal/TaskCreate.scala

```diff
@@ -184,44 +166,85 @@ private[eval] object TaskCreate {
     Async(start, trampolineBefore = false, trampolineAfter = false)
   }
 
-  /**
-    * Creates a callback that will forward calls to the provided one
-    * and call it on the default Scheduler (from Context).
-    *
-    * Useful in case the `Callback` could be called from unknown
-    * thread pools.
-    */
-  private def executeCallbackOn[A](ctx: Context, cb: Callback[Throwable, A]): Callback[Throwable, A] = {
-    new Callback[Throwable, A] {
-      override def onSuccess(value: A): Unit =
-        try {
-          ctx.scheduler.execute(new Runnable {
-            def run(): Unit = {
-              ctx.frameRef.reset()
-              cb.onSuccess(value)
-            }
-          })
-        } catch {
-          case e: RejectedExecutionException =>
-            Callback
-              .trampolined(cb)(ctx.scheduler)
-              .onError(e)
-        }
-
-      override def onError(e: Throwable): Unit =
-        try {
-          ctx.scheduler.execute(new Runnable {
-            def run(): Unit = {
-              ctx.frameRef.reset()
-              cb.onError(e)
-            }
-          })
-        } catch {
-          case e: RejectedExecutionException =>
-            Callback
-              .trampolined(cb)(ctx.scheduler)
-              .onError(e)
-        }
+  private[internal] def protectedCallback[A](
+    ctx: Context,
+    shouldPop: Boolean,
+    cb: Callback[Throwable, A]): Callback[Throwable, A] =
+    new CallbackForCreate[A](ctx, shouldPop, cb)
+
+  private final class CallbackForCreate[A](ctx: Context, threadId: Long, shouldPop: Boolean, cb: Callback[Throwable, A])
+    extends Callback[Throwable, A] with TrampolinedRunnable {
+
+    private[this] val state = AtomicInt(0)
+    private[this] var value: A = _
+    private[this] var error: Throwable = _
+    private[this] var isSameThread = false
+
+    def this(ctx: Context, shouldPop: Boolean, cb: Callback[Throwable, A]) =
+      this(ctx, Platform.currentThreadId(), shouldPop, cb)
+
+    override def onSuccess(value: A): Unit = {
+      if (state.compareAndSet(0, 1)) {
+        this.value = value
+        startExecution()
+      } else {
+        throw new IllegalStateException("Callback.onSuccess signaled multiple times")
+      }
+    }
+
+    override def onError(e: Throwable): Unit = {
+      if (state.compareAndSet(0, 2)) {
+        this.error = e
+        startExecution()
+      } else {
+        throw new IllegalStateException("Callback.onSuccess onError multiple times", e)
+      }
+    }
+
+    private def startExecution(): Unit = {
+      // Optimization — if the callback was called on the same thread
+      // where it was created, then we are not going to fork
+      isSameThread = Platform.currentThreadId() == threadId
+      try {
+        ctx.scheduler.execute(
+          if (isSameThread)
+            this
```

Comment:
if it's the same thread, do we really need an async boundary from `execute` here? Can't we just call `run()`?

Reply:
Not really, because we'd run into stack safety issues. Some sort of async boundary, either via a real fork or via the trampoline, is needed.

Before this change the callback was wrapped in a `Callback.trampoline()`. It's also why we have the settings `trampolineBefore` and `trampolineAfter` on `Async`. The point of those settings was to avoid wrapping callbacks in additional instances and we don't need to use those here, since we are wrapping the callback anyway.